pkg/queue/sharedmain/main.go

```diff
@@ -246,29 +239,39 @@ func Main(opts ...Option) error {
 	// Enable TLS when certificate is mounted.
 	tlsEnabled := exists(logger, certPath) && exists(logger, keyPath)
 
-	mainServer, drainer := buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, false)
+	servingHandler, drainer := servingHandler(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint)
+	adminHandler := adminHandler(d.Ctx, logger, drainer)
+
+	// Enable TLS server when activator server certs are mounted.
+	// At this moment activator with TLS does not disable HTTP.
+	// See also https://github.com/knative/serving/issues/12808.
 	httpServers := map[string]*http.Server{
-		"main":    mainServer,
-		"metrics": buildMetricsServer(protoStatReporter),
-		"admin":   buildAdminServer(d.Ctx, logger, drainer),
+		"main": servingServer(":"+env.QueueServingPort, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			logger.Info("http")
+			servingHandler.ServeHTTP(w, r)
+		}),
+		),
+		"admin":   adminServer(":"+strconv.Itoa(networking.QueueAdminPort), adminHandler),
+		"metrics": metricsServer(protoStatReporter),
 	}
+
 	if env.EnableProfiling {
 		httpServers["profile"] = profiling.NewServer(profiling.NewHandler(logger, true))
 	}
 
-	// Enable TLS server when activator server certs are mounted.
-	// At this moment activator with TLS does not disable HTTP.
-	// See also https://github.com/knative/serving/issues/12808.
-	var tlsServers map[string]*http.Server
+	tlsServers := map[string]*http.Server{
+		"main": servingServer(":"+env.QueueServingTLSPort, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			logger.Info("tls")
+			servingHandler.ServeHTTP(w, r)
+		}),
+		),
+		"admin": adminServer(":"+strconv.Itoa(networking.QueueAdminPort), adminHandler),
+	}
+
 	if tlsEnabled {
-		mainTLSServer, drainer := buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, true /* enable TLS */)
-		tlsServers = map[string]*http.Server{
-			"tlsMain":  mainTLSServer,
-			"tlsAdmin": buildAdminServer(d.Ctx, logger, drainer),
-		}
-		// Drop admin http server as we Use TLS for the admin server.
-		// TODO: The drain created with mainServer above is lost. Unify the two drain.
 		delete(httpServers, "admin")
```

Comment:
Can we keep the comment `// Drop admin http server as we Use TLS for the admin server.`? I think it's not super obvious why the admin server is dropped here.